Sources/PackageFingerprint/FilePackageFingerprintStorage.swift

```diff
@@ -0,0 +1,190 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import Dispatch
+import Foundation
+import PackageModel
+import TSCBasic
+import TSCUtility
+
+public struct FilePackageFingerprintStorage: PackageFingerprintStorage {
+    let fileSystem: FileSystem
+    let directory: AbsolutePath
+
+    private let encoder: JSONEncoder
+    private let decoder: JSONDecoder
+
+    init(customFileSystem: FileSystem? = nil, customDirectory: AbsolutePath? = nil) {
+        self.fileSystem = customFileSystem ?? localFileSystem
+        self.directory = customDirectory ?? self.fileSystem.dotSwiftPM.appending(component: "fingerprints")
```

Comment:
the default location can also be encoded into Workspace.Location / DefaultLocation which handles similar set ups

Reply:
Making both required so no need to have defaults here.